Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift

```diff
@@ -18,46 +18,85 @@ import SwiftTreeSitter
         "CodeEdit.TreeSitterClient"
     }
 
-    internal var parser: Parser
-    internal var tree: Tree?
-    internal var languageQuery: Query?
-
-    private var textProvider: ResolvingQueryCursor.TextProvider
+    internal var primaryLayer: TreeSitterLanguage
+    internal var layers: [LanguageLayer] = []
 
-    /// The queue to do  tree-sitter work on for large edits/queries
-    private let queue: DispatchQueue = DispatchQueue(label: "CodeEdit.CodeEditTextView.TreeSitter",
-                                                     qos: .userInteractive)
+    internal var textProvider: ResolvingQueryCursor.TextProvider
 
     /// Used to ensure safe use of the shared tree-sitter tree state in different sync/async contexts.
-    private let semaphore: DispatchSemaphore = DispatchSemaphore(value: 1)
+    internal let semaphore: DispatchSemaphore = DispatchSemaphore(value: 1)
+
+    internal enum Constants {
+        /// The maximum amount of limits a cursor can match during a query.
+        /// Used to ensure performance in large files, even though we generally limit the query to the visible range.
+        /// Neovim encountered this issue and uses 64 for their limit. Helix uses 256 due to issues with some
+        /// languages when using 64.
+        /// See: https://github.com/neovim/neovim/issues/14897
+        /// And: https://github.com/helix-editor/helix/pull/4830
+        static let treeSitterMatchLimit = 256
+    }
+
+    // MARK: - Init/Config
 
     /// Initializes the `TreeSitterClient` with the given parameters.
     /// - Parameters:
     ///   - codeLanguage: The language to set up the parser with.
     ///   - textProvider: The text provider callback to read any text.
-    public init(codeLanguage: CodeLanguage, textProvider: @escaping ResolvingQueryCursor.TextProvider) throws {
-        parser = Parser()
-        languageQuery = TreeSitterModel.shared.query(for: codeLanguage.id)
-        tree = nil
-
+    public init(codeLanguage: CodeLanguage, textProvider: @escaping ResolvingQueryCursor.TextProvider) {
         self.textProvider = textProvider
+        self.primaryLayer = codeLanguage.id
+        setLanguage(codeLanguage: codeLanguage)
+    }
+
+    /// Sets the primary language for the client. Will reset all layers, will not do any parsing work.
+    /// - Parameter codeLanguage: The new primary language.
+    public func setLanguage(codeLanguage: CodeLanguage) {
+        // Remove all trees and languages, everything needs to be re-parsed.
+        layers.removeAll()
+
+        primaryLayer = codeLanguage.id
+        layers = [
+            LanguageLayer(id: codeLanguage.id,
+                          parser: Parser(),
+                          supportsInjections: codeLanguage.additionalHighlights?.contains("injections") ?? false,
+                          tree: nil,
+                          languageQuery: TreeSitterModel.shared.query(for: codeLanguage.id),
+                          ranges: [])
```

Comment:
### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
            LanguageLayer(
                id: codeLanguage.id,
                parser: Parser(),
                supportsInjections: codeLanguage.additionalHighlights?.contains("injections") ?? false,
                tree: nil,
                languageQuery: TreeSitterModel.shared.query(for: codeLanguage.id),
                ranges: []
            )
```